<a href="https://colab.research.google.com/github/madhan444-s/Madhan_INFO5731_Spring2024/blob/main/Dadi_Madhan_Assignment_01_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment One**

This assignment includes range of programming tasks and data analysis exercises, aiming to enhance your Python skills and proficiency in string manipulation, numerical computations, and data science. Follow the instructions carefully and design well-structured Python programs to address each question.

Late Submission will have a penalty of 10% reduction for each day after the deadline.

# Question 1 (10 points).

The formula that describes the volume V of a hemisphere with radius r is the following:

$ V=\frac{2}{3}\ast\ \pi\ast\ r^3 $

Write a program to calculate the value of V when r is in the range of even numbers from 1-20. Output the result in the following format:

	r	V
	2	…
	4	…
	6	…
	.
	.
	.
	20	…


In [1]:
# Your code here
import math

def calculateHemisphereVolume(radius):
    return (2/3) * math.pi * radius**3

def main():
    print("r\tV")
    for r in range(2, 21, 2):
        v = calculateHemisphereVolume(r)
        print(f"{r}\t{v:.2f}")

if __name__ == "__main__":
    main()


r	V
2	16.76
4	134.04
6	452.39
8	1072.33
10	2094.40
12	3619.11
14	5747.02
16	8578.64
18	12214.51
20	16755.16


# Question 2 (20 points)

(a) Write a Python program to generate a sequence of 12 random numbers falling within the range of 50 to 150.

(b) Arrange the generated numbers in ascending order. Display the sorted list.

(c) Extend the program to arrange the numbers in descending order as well. Present the sorted list accordingly.

(d) Compute and output the average of the entire group of generated numbers.

(e) Enhance the program to calculate and display the average of the first 6 numbers from both the ascending and descending sorted lists.

In [9]:
# Write code for each of the sub parts with proper comments.

import random

# Function to generate a sequence of 12 random numbers in the range of 50 to 150
def generateRandomNumbers():
    return [random.randint(50, 150) for _ in range(12)]

# Function to calculate and return the average of a list of numbers
def calculateAverage(numbers):
    return sum(numbers) / len(numbers)

def main():
    # (a) Generate a sequence of 12 random numbers
    randomNumbers = generateRandomNumbers()

    # (b) Arrange the generated numbers in ascending order and display the sorted list
    ascendingSort = sorted(randomNumbers)
    print("Ascending Sorted List:", ascendingSort)

    # (c) Arrange the numbers in descending order and display the sorted list
    descendingSort = sorted(randomNumbers, reverse=True)
    print("Descending Sorted List:", descendingSort)

    # (d) Compute and output the average of the entire group of generated numbers
    average_all_numbers = calculateAverage(randomNumbers)
    print(f"Average of all numbers: {average_all_numbers:.2f}")

    # (e) Calculate and display the average of the first 6 numbers from both lists
    averageFirst6Ascending = calculateAverage(ascendingSort[:6])
    averageFirst6Descending = calculateAverage(descendingSort[:6])

    print(f"Average of the first 6 numbers (Ascending): {averageFirst6Ascending:.2f}")
    print(f"Average of the first 6 numbers (Descending): {averageFirst6Descending:.2f}")

if __name__ == "__main__":
    main()


Ascending Sorted List: [54, 74, 74, 89, 93, 103, 106, 110, 134, 139, 141, 143]
Descending Sorted List: [143, 141, 139, 134, 110, 106, 103, 93, 89, 74, 74, 54]
Average of all numbers: 105.00
Average of the first 6 numbers (Ascending): 81.17
Average of the first 6 numbers (Descending): 128.83


# Question 3 (20 points).

Create a Python program that encompasses various string manipulation task:

(a) Prompt the user to input two sentences.

(b) Transform each sentence into a list, where each element corresponds to a word in the sentence. For example, if the sentences are "Hello, how are you doing today?" and "This is an example sentence for the program!"

The program should output:
['Hello', ',', 'how', 'are', 'you', 'doing', 'today', '?', 'This', 'is', 'an', 'example', 'sentence', 'for', 'the', 'program', '!']

(c) Compare the two sentences. Print out a list of words that occur in both sentences.

(d) Remove any punctuation from the sentence and print the sanitized sentences. It should remove at least the following characters, but it can remove more: period(“.”), comma (“,”), semicolon (“;”), and colon (“:”).

(e) Count the number of vowels in each of the sentences. Print out each vowel and the occurrences  in the each of the two sentences separately, such as:

Sentence1: a:2, e:2, i:0, etc.
Sentence2: a:1, e:2, i:2, etc.

Design a well-structured Python program to execute these tasks efficiently.

In [2]:
import string

def getWordList(sentence):
    # Transform a sentence into a list of words
    words = sentence.split()

    return words

def commonWords(list1, list2):
    # Find common words between two lists
    return list(set(list1) & set(list2))

def removePunctuation(sentence):
    # Remove specified punctuation from the sentence
    translator = str.maketrans('', '', string.punctuation)
    return sentence.translate(translator)

def countVowels(sentence):
    # Count occurrences of each vowel in a sentence
    vowels = "aeiou"
    vowelCount = {vowel: sentence.lower().count(vowel) for vowel in vowels}
    return vowelCount

def printVowelCount(vowelCount, sentenceNumber):
    print(f"Sentence{sentenceNumber}: ", end="")
    for vowel, count in vowelCount.items():
        print(f"{vowel}:{count}", end=", ")
    print()

def main():
    # Prompt user for input
    sentence1 = input("Enter the first sentence: ")
    sentence2 = input("Enter the second sentence: ")

    # Transform each sentence into a list of words
    words1 = getWordList(sentence1)
    words2 = getWordList(sentence2)

    # Print the combined list of words from both sentences
    combinedWords = words1 + words2
    print(combinedWords)

    # Print common words in both sentences
    common = commonWords(words1, words2)
    print("Common words:", common)

    # Remove punctuation from sentences
    sanitizedSentence1 = removePunctuation(sentence1)
    sanitizedSentence2 = removePunctuation(sentence2)
    print("Sanitized Sentence 1:", sanitizedSentence1)
    print("Sanitized Sentence 2:", sanitizedSentence2)

    # Count vowels in each sentence
    vowelCount1 = countVowels(sentence1)
    vowelCount2 = countVowels(sentence2)

    # Print vowel counts
    print("\nVowel Count:")
    printVowelCount(vowelCount1, 1)
    printVowelCount(vowelCount2, 2)

if __name__ == "__main__":
    main()


Enter the first sentence: Hello, how are you doing today?
Enter the second sentence: This is an example sentence for the program!
['Hello,', 'how', 'are', 'you', 'doing', 'today?', 'This', 'is', 'an', 'example', 'sentence', 'for', 'the', 'program!']
Common words: []
Sanitized Sentence 1: Hello how are you doing today
Sanitized Sentence 2: This is an example sentence for the program

Vowel Count:
Sentence1: a:2, e:2, i:1, o:5, u:1, 
Sentence2: a:3, e:6, i:2, o:2, u:0, 


# Question 4 (20 Points)

Design and implement a simple command-line calculator program. The program should handle both integers and floating-point numbers.

(a) Write program that takes a mathematical expression as input and evaluates it. The program should support addition, subtraction, multiplication, and division.

(b) Extend the calculator to handle parentheses in mathematical expressions. Implement the necessary logic to evaluate expressions with nested parentheses.

(c) Enhance your calculator to include additional mathematical operations such as exponentiation, square root, and modulus.

(d) Implement error handling in the calculator program. Handle cases such as division by zero and invalid input expressions. Provide meaningful error messages.

(e) Implement a command to exit the calculator.

In [8]:
# Your code here
import re
import math

def evaluateExpression(expression):
    try:
        result = eval(expression)
        return result
    except ZeroDivisionError:
        return "Error: Division by zero"
    except SyntaxError:
        return "Error: Invalid expression"

def handleCalc():
    while True:
        userInput = input("Enter a mathematical expression (type 'exit' to quit): ")

        if userInput.lower() == 'exit':
            print("Exiting calculator. Goodbye!")
            break

        # Check for valid input
        if not re.match(r'^[0-9+\-*/().^%sqrt\s]+$', userInput):
            print("Error: Invalid characters in expression")
            continue

        result = evaluateExpression(userInput)
        print("Result:", result)

if __name__ == "__main__":
    print("Simple Command-Line Calculator")
    print("Supported operations: + - * / ^ % sqrt")
    print("Type 'exit' to quit")

    handleCalc()


Simple Command-Line Calculator
Supported operations: + - * / ^ % sqrt
Type 'exit' to quit
Enter a mathematical expression (type 'exit' to quit): 2+3
Result: 5
Enter a mathematical expression (type 'exit' to quit): (2+3)*5
Result: 25
Enter a mathematical expression (type 'exit' to quit): a+b
Error: Invalid characters in expression
Enter a mathematical expression (type 'exit' to quit): exit
Exiting calculator. Goodbye!


# Question 5 (30 points).

In the field of Data Science, data is often formatted as a comma-delimited (CSV) file, where each line in a file corresponds to a field's value. Refer to Chapter 9 (p. 106) for more information on comma-delimited files.

To answer this question, you will need to download the file [Assignment1_denton_housing.csv](https://raw.githubusercontent.com/unt-iialab/info5731-spring2022/main/assignments/Assignment1_denton_housing.csv) included with this assignment. The file contains statistics about housing in the city of Denton from 2008 to 2014. Write a program to implement the following questions.

(a) (5 pts) Calculate the difference in the number of Occupied Housing Units from year to year and print it. The difference must be calculated for the consecutive years such as 2008-2009, 2009-2010 etc. Finally, print the values in the ascending order.

(b) (5 pts) For all the years, calculate the percentage of housing units which are vacant an occupied. Print the results in the following format:

Year Vacant Housing Units Occupied Housing Units

    2008  30%   70%                          
	2009 ----- -----
	2010 ----- -----
    2011 ----- -----
	2012 ----- -----
    2013 ----- -----
	2014 ----- -----
  
  

(c) (10 pts) Calculate and print the valued and years in which the highest number of housing units were vacant and occupied. Print the results in the following format:  

                              Year	Value
	Vacant Housing Units	  -----   -----
	Occupied Housing Units	-----   -----

(d) (10 pts) Calculate the harmonic mean of the total housing units and print it out. You can find the information about harmonic mean here: https://ncalculators.com/statistics/harmonic-mean-calculator.htm

In [6]:
import pandas as pd

# Read the CSV file
df = pd.read_csv("Assignment1_denton_housing.csv")

# Calculate the difference in the number of Occupied Housing Units from year to year and print it in ascending order
df_diff = df[df['title_field'] == 'Occupied Housing Units'].copy()
df_diff['Difference'] = df_diff['value'].diff()
df_diff_sorted = df_diff[['year', 'Difference']].dropna().sort_values(by='Difference', ascending=True)
print("Year | Difference in Occupied Housing Units")
print(df_diff_sorted.to_string(index=False))

# Calculate the percentage of vacant and occupied housing units for each year and print the results
df_pivot = df.pivot(index='year', columns='title_field', values='value')
df_pivot['Total Housing Units'] = df_pivot['Vacant Housing Units'] + df_pivot['Occupied Housing Units']
df_pivot['Vacant Housing Units (%)'] = (df_pivot['Vacant Housing Units'] / df_pivot['Total Housing Units']) * 100
df_pivot['Occupied Housing Units (%)'] = (df_pivot['Occupied Housing Units'] / df_pivot['Total Housing Units']) * 100
df_percentages = df_pivot[['Vacant Housing Units (%)', 'Occupied Housing Units (%)']]
print("\nYear | Vacant Housing Units (%) | Occupied Housing Units (%)")
print(df_percentages.to_string())

# Calculate and print the years in which the highest number of housing units were vacant and occupied
max_vacant_year = df_pivot['Vacant Housing Units'].idxmax()
max_vacant_value = df_pivot['Vacant Housing Units'].max()
max_occupied_year = df_pivot['Occupied Housing Units'].idxmax()
max_occupied_value = df_pivot['Occupied Housing Units'].max()

print("\nYear | Value")
print(f"Vacant Housing Units | {max_vacant_year} | {max_vacant_value}")
print(f"Occupied Housing Units | {max_occupied_year} | {max_occupied_value}")

Year | Difference in Occupied Housing Units
 year  Difference
 2009     -5091.0
 2013     -2622.0
 2012     -1565.0
 2011     -1404.0
 2010       303.0
 2008       795.0

Year | Vacant Housing Units (%) | Occupied Housing Units (%)
title_field  Vacant Housing Units (%)  Occupied Housing Units (%)
year                                                             
2008                        10.882653                   89.117347
2009                         9.071116                   90.928884
2010                         8.142556                   91.857444
2011                         9.379522                   90.620478
2012                         6.677600                   93.322400
2013                         8.801788                   91.198212
2014                         5.730111                   94.269889

Year | Value
Vacant Housing Units | 2008 | 4483
Occupied Housing Units | 2014 | 46295


# Mandatory Question

Provide your thoughts on the assignment. What did you find challenging, and what aspects did you enjoy? Your opinion on the provided time to complete the assignment. Share any suggestions for improvement. Your feedback will be valuable for enhancing the learning experience.

In [ ]:
# Write your response below
"""The assignment provided is well-structured and covers various aspects of
data analysis and manipulation, which are commonly encountered in the field of data science.

Challenging aspects
Performing calculations such as finding the difference between housing
units over consecutive years, calculating percentages, and identifying maximum values
require careful data manipulation and logic implementation.

Enjoyable aspects:
Designing solutions to the given tasks provided an opportunity to apply various programming and analytical
skills to real-world problems.
Exploring different aspects of data analysis, such as calculating percentages, identifying trends,
and handling data errors, provided valuable learning experiences.

Suggestions for improvement
Providing additional context or background information about the dataset and its significance to the problem domain
can enhance the learning experience.*/
"""